org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/XtextProjectHelper.java

```diff
@@ -24,7 +24,7 @@
 
 	public static boolean hasNature(IProject project) {
```

Comment:
Where did the project come from? E.g. which piece of code called hasNature with a null argument? Usually projects are obtained from the workspace and are not null.

Reply:
We are working on containers with external files. In that specific case Storage2UriMapperImpl returns a collection of mappings from a storage to a null project. This causes the NPE in org.eclipse.xtext.builder.impl.QueuedBuildData.queueURI(URI) when the helper is invoked with a null project.
As said, the project itself is buggy ATM, but the NPE here might not be the right answer also.

Reply:
It depends. An exception helped you to easily spot that you called this method transitively with a null project. A simple false wouldn't have helped, I suppose. I'm not convinced.

Reply:
Then Preconditions#checkNotNull might worth consideration. It effectively also throws a NPE, but from the stack one could see that this is actually a guarded check.

Reply:
Doesn't make any difference at all at this exact line of code where we dereference the argument and only the argument immediately and where we do not store the argument in a field. But please go ahead and do that.